vllm/v1/core/sched/waiting_queue.py

```diff
@@ -0,0 +1,166 @@
+        self.appendleft(request)
+
+    def extend_left_requests(self, requests: WaitingQueue) -> None:
+        """Extend left with requests from another WaitingQueue."""
+        self.extendleft(reversed(list(requests)))
+
+    def remove_request(self, request: Request) -> None:
+        """Remove a specific request from the queue."""
+        self.remove(request)
+
+    def __bool__(self) -> bool:
+        """Check if queue has any requests."""
+        return len(self) > 0
+
+    def __len__(self) -> int:
+        """Get number of requests in queue."""
+        return super().__len__()
+
+    def __iter__(self) -> Iterator[Request]:
+        """Iterate over the queue according to FCFS policy."""
+        return super().__iter__()
+
+
+class PriorityWaitingQueue(WaitingQueue):
+    """A priority queue that supports heap operations."""
+
+    def __init__(self) -> None:
+        self._heap: list[tuple[int, float, Request]] = []
+
+    def add_request(self, request: Request) -> None:
+        """Add a request to the queue according to priority policy."""
+        heapq.heappush(self._heap,
+                       (request.priority, request.arrival_time, request))
+
+    def pop_request(self) -> Request:
+        """Pop a request from the queue according to priority policy."""
+        if not self._heap:
+            raise IndexError("pop from empty heap")
+        _, _, request = heapq.heappop(self._heap)
+        return request
+
+    def push_request(self,
+                     request: Request,
+                     priority: int = 0,
+                     arrival_time: float = 0.0) -> None:
+        """Push a request back to the queue (used for skipped requests)."""
+        heapq.heappush(self._heap, (priority, arrival_time, request))
+
+    def extend_left_requests(self, requests: WaitingQueue) -> None:
+        """Extend left with requests from another WaitingQueue."""
+        if isinstance(requests, PriorityWaitingQueue):
+            for item in requests._heap:
+                heapq.heappush(self._heap, item)
+        else:
+            # Handle FCFS queue case
+            for request in requests:
+                heapq.heappush(
+                    self._heap,
+                    (request.priority, request.arrival_time, request))
+
+    def remove_request(self, request: Request) -> None:
+        """Remove a specific request from the queue."""
+        self._heap = [(p, t, r) for p, t, r in self._heap if r != request]
+        heapq.heapify(self._heap)
+
+    def __bool__(self) -> bool:
+        """Check if queue has any requests."""
+        return bool(self._heap)
+
+    def __len__(self) -> int:
+        """Get number of requests in queue."""
+        return len(self._heap)
+
+    def __iter__(self) -> Iterator[Request]:
+        """Iterate over the queue according to priority policy."""
+        heap_copy = self._heap[:]
+        while heap_copy:
+            _, _, request = heapq.heappop(heap_copy)
+            yield request
```

Comment:
QQ: Where is this used? Also, should we iterate with the strict order? This seems pretty expensive.